runners/core-java/src/main/java/org/apache/beam/runners/core/WatermarkHold.java

```diff
@@ -115,11 +115,15 @@ private Instant shift(Instant timestamp, W window) {
         windowingStrategy
             .getTimestampCombiner()
             .assign(window, windowingStrategy.getWindowFn().getOutputTime(timestamp, window));
-    checkState(!shifted.isBefore(timestamp),
-        "TimestampCombiner moved element from %s to earlier time %s for window %s",
-        BoundedWindow.formatTimestamp(timestamp),
-        BoundedWindow.formatTimestamp(shifted),
-        window);
+    // check if shifted is before timestamp first, before calling checkState()
+    // to avoid calling BoundedWindow.formatTimestamp() every time
+    if (shifted.isBefore(timestamp)) {
+      checkState(false,
```

Comment:
checkState doesn't call String.format(). Shouldn't I call checkState instead to make it consistent with the other calls to checkState?

Reply:
checkState calls Strings.lenientFormat internally https://github.com/google/guava/blob/0cd4e9faa1360da4a343f84cb275d6eda0c5e732/guava/src/com/google/common/base/Preconditions.java#L531

Strings.lenientFormat javadoc says to use String.format instead:
https://github.com/google/guava/blob/0cd4e9faa1360da4a343f84cb275d6eda0c5e732/guava/src/com/google/common/base/Strings.java#L232

Reply:
Agree with Luke. `checkState` is just a tiny readability wrapper when it can be used. So if you have to do an `if` then you might as well just throw the exception directly.

Reply:
I see. Fixed. Thanks for your comments!